source/lac/petsc_matrix_base.cc

```diff
@@ -605,7 +605,12 @@ namespace PETScWrappers
   void
   MatrixBase::transpose()
   {
+#  if DEAL_II_PETSC_VERSION_LT(3, 8, 0)
     const PetscErrorCode ierr = MatTranspose(matrix, MAT_REUSE_MATRIX, &matrix);
+#  else
+    const PetscErrorCode ierr =
+      MatTranspose(matrix, MAT_INPLACE_MATRIX, &matrix);
```

Comment:
Can you add a comment that explains what the difference between `MAT_REUSE|INPLACE_MATRIX` is? And is it really necessary to keep the old code around, or do 3.8 and before really require one flag and 3.9 and later require another?

Reply:
@bangerth I do not know.

The PETSc changelog states [1]: "MatTranspose() now properly handles MAT_INITIAL_MATRIX, MAT_REUSE_MATRIX, MAT_INPLACE_MATRIX."

I do not have a PETSc-3.7 version available any more to properly test the code, thus I introduced the same preprocessor switch that was done in #5655. 

[1] https://www.mcs.anl.gov/petsc/documentation/changes/38.html

Reply:
OK. Who knows. Let's just go with what you have then!

Reply:
The PETSc documentation says for the enum:
```
- MAT_INITIAL_MATRIX - create a new matrix
- MAT_REUSE_MATRIX - reuse the matrix created with a previous call that used MAT_INITIAL_MATRIX
- MAT_INPLACE_MATRIX - replace the first input matrix with the new matrix (not applicable to all functions)
- MAT_IGNORE_MATRIX - do not create a new matrix or reuse a give matrix, just ignore that matrix argument (not applicable to all functions)
```
and for `MatTranspose`:
```
If you use MAT_INPLACE_MATRIX then you must pass in &mat for B.
MAT_REUSE_MATRIX causes the B matrix from a previous call to this function with MAT_INITIAL_MATRIX to be used.
```
If I understand this correctly, `MAT_REUSE_MATRIX` should not work because we didn't call the function with `MAT_INITIAL_MATRIX` before.
Unfortunately, there is no PETSc example that uses `MAT_REUSE_MATRIX`.

I guess we should just test with an older `PETSc` version once we have a test case.

Reply:
@masterleinad Agreed. The question is what old PETSc versions did (accidentally interpreting `MAT_REUSE_MATRIX` as `MAT_INPLACE_MATRIX`, or similar) and what parameter we have to supply to get the intented behavior.

Reply:
At least, `PETSc 3.6.3` doesn't know about `MAT_INPLACE_MATRIX`.
3.7 is the first version knowing about this enum.